Cache full script execution results in addition to signatures

TheBlueMatt opened the pull request:
This adds a new CuckooCache in validation, caching whether all of a
transaction's scripts were valid with a given set of script flags.

Unlike previous attempts at caching an entire transaction's
validity, which have nearly universally introduced consensus
failures, this only caches the validity of a transaction's
scriptSigs. As these are pure functions of the transaction and
data it commits to, this should be much safer.

This is somewhat duplicative with the sigcache, as entries in the
new cache will also have several entries in the sigcache. However,
the sigcache is kept both as ATMP relies on it and because it
prevents malleability-based DoS attacks on the new higher-level
cache. Instead, the -sigcachesize option is re-used - cutting the
sigcache size in half and using the newly freed memory for the
script execution cache.

Transactions which match the script execution cache never even have
entries in the script check thread's workqueue created.

Note that the cache is indexed only on the script execution flags
and the transaction's witness hash. While this is sufficient to
make the CScriptCheck() calls pure functions, this introduces
dependancies on the mempool calculating things such as the
PrecomputedTransactionData object, filling the CCoinsViewCache, etc
in the exact same way as ConnectBlock. I belive this is a reasonable
assumption, but should be noted carefully.

In a rather naive benchmark (reindex-chainstate up to block 284k
with cuckoocache always returning true for contains(),
-assumevalid=0 and a very large dbcache), this connected blocks
~1.7x faster.

TheBlueMatt committed:
Pull script verify flags calculation out of ConnectBlock

TheBlueMatt committed:
Cache full script execution results in addition to signatures

This adds a new CuckooCache in validation, caching whether all of a
transaction's scripts were valid with a given set of script flags.

Unlike previous attempts at caching an entire transaction's
validity, which have nearly universally introduced consensus
failures, this only caches the validity of a transaction's
scriptSigs. As these are pure functions of the transaction and
data it commits to, this should be much safer.

This is somewhat duplicative with the sigcache, as entries in the
new cache will also have several entries in the sigcache. However,
the sigcache is kept both as ATMP relies on it and because it
prevents malleability-based DoS attacks on the new higher-level
cache. Instead, the -sigcachesize option is re-used - cutting the
sigcache size in half and using the newly freed memory for the
script execution cache.

Transactions which match the script execution cache never even have
entries in the script check thread's workqueue created.

Note that the cache is indexed only on the script execution flags
and the transaction's witness hash. While this is sufficient to
make the CScriptCheck() calls pure functions, this introduces
dependancies on the mempool calculating things such as the
PrecomputedTransactionData object, filling the CCoinsViewCache, etc
in the exact same way as ConnectBlock. I belive this is a reasonable
assumption, but should be noted carefully.

In a rather naive benchmark (reindex-chainstate up to block 284k
with cuckoocache always returning true for contains(),
-assumevalid=0 and a very large dbcache), this connected blocks
~1.7x faster.

sipa commented on src/validation.cpp:
```diff
@@ -1356,7 +1373,32 @@ bool CheckTxInputs(const CTransaction& tx, CValidationState& state, const CCoins
+void InitScriptExecutionCache() {
+    // nMaxCacheSize is unsigned. If -maxsigcachesize is set to zero,
+    // setup_bytes creates the minimum possible cache (2 elements).
+    size_t nMaxCacheSize = std::min(std::max((int64_t)0, GetArg("-maxsigcachesize", DEFAULT_MAX_SIG_CACHE_SIZE) / 2), MAX_MAX_SIG_CACHE_SIZE) * ((size_t) 1 << 20);
```
I think the division should be outside of GetArg. Otherwise, if you specify -maxsigcachesize=32, you end up with a total of 64MiB worth of caches.

gmaxwell replied:
Good, because the division is outside of the GetArg. :)

sipa replied:
It seems I am blind.

TheBlueMatt commented:
Fixed test_bitcoin segfaulting as it didnt init the script cache as it does the sigcache.

gmaxwell commented:
Concept ACK.

JeremyRubin reviewed (changes requested):
Concept Ack!

instagibbs commented on src/validation.cpp:
```diff
@@ -1378,13 +1427,26 @@ bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsVi
+            // correct (ie that the transaction hash which is in tx's prevouts
+            // properly commits to the scriptPubKey in the inputs view of that
+            // transaction).
+            static uint256 nonce(GetRandHash());
```
Struggling to understand how a unique nonce per cache entry vs per cache works.

JeremyRubin replied:
it's static so it is per cache.

instagibbs commented:
concept ACK

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -12,6 +12,7 @@
 #include "consensus/consensus.h"
 #include "consensus/merkle.h"
 #include "consensus/validation.h"
+#include "cuckoocache.h"
```
Is using the public interface of CuckooCache::cache really using its "internals"?

JeremyRubin replied:
I guess not.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -1356,7 +1382,32 @@ bool CheckTxInputs(const CTransaction& tx, CValidationState& state, const CCoins
 
-bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsViewCache &inputs, bool fScriptChecks, unsigned int flags, bool cacheStore, PrecomputedTransactionData& txdata, std::vector<CScriptCheck> *pvChecks)
+namespace {
+class ShaHashSplitter {
```
Rebased on that, removed this class.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -1376,13 +1427,26 @@ bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsVi
+            // correct (ie that the transaction hash which is in tx's prevouts
+            // properly commits to the scriptPubKey in the inputs view of that
+            // transaction).
+            static uint256 nonce(GetRandHash());
```
I'd much much rather have a static in the function than add yet another class to compile. We need to have fewer two-line wrapper classes, not more...our memory usage is already insane.

JeremyRubin replied:
Memory usage during compile time? Is it really that bad for an added class?

Can you at least move the static decl to the top of the function?

TheBlueMatt replied:
Done. Moved the static to the top of the function with the static CuckooCache as well.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -1376,13 +1427,26 @@ bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsVi
+            // transaction).
+            static uint256 nonce(GetRandHash());
+            uint256 hashCacheEntry;
+            CSHA256().Write(nonce.begin(), 32).Write(tx.GetWitnessHash().begin(), 32).Write((unsigned char*)&flags, sizeof(flags)).Finalize(hashCacheEntry.begin());
```
Using 19 bytes of nonce, thats way more than enough.

TheBlueMatt commented:
Addressed Jeremy's comments aside from the request for a wrapper class, I think we need fewer dummy classes, not more :/.

gmaxwell commented on src/validation.cpp:
```diff
@@ -1376,13 +1401,28 @@ bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsVi
+            static uint256 nonce(GetRandHash());
+            uint256 hashCacheEntry;
+            // We only use the first 19 bytes of nonce to avoid a second SHA
+            // round - giving us 19 + 32 + 4 = 55 bytes (+ 8 + 1 = 64)
```
Static assert on the size of the flags so the nonce gets reduced if the flags are made 64-bits in the future?

TheBlueMatt replied:
Done.

sdaftuar commented:
> Note that the cache is indexed only on the script execution flags
and the transaction's witness hash. While this is sufficient to
make the CScriptCheck() calls pure functions, this introduces
dependancies on the mempool calculating things such as the
PrecomputedTransactionData object, filling the CCoinsViewCache, etc
in the exact same way as ConnectBlock. I belive this is a reasonable
assumption, but should be noted carefully.

So if I understand correctly, this makes CCoinsViewMempool, and therefore the mempool itself, consensus critical, which I would prefer to avoid.  Would we still get a performance benefit if we were to include the coins being spent in the hash for the index lookup?

TheBlueMatt commented:
@sdaftuar I'm not convinced its a massive concern, but I went ahead and added a wrapper which checks each scriptPubKey returned by the CCoinsViewCache is the one committed to by the input's prevout hash, which I believe removes that dependancy entirely.

JeremyRubin commented on src/script/sigcache.cpp:
```diff
@@ -74,7 +74,7 @@ void InitSignatureCache()
-    size_t nMaxCacheSize = std::min(std::max((int64_t)0, GetArg("-maxsigcachesize", DEFAULT_MAX_SIG_CACHE_SIZE)), MAX_MAX_SIG_CACHE_SIZE) * ((size_t) 1 << 20);
+    size_t nMaxCacheSize = std::min(std::max((int64_t)0, GetArg("-maxsigcachesize", DEFAULT_MAX_SIG_CACHE_SIZE) / 2), MAX_MAX_SIG_CACHE_SIZE) * ((size_t) 1 << 20);
     size_t nElems = signatureCache.setup_bytes(nMaxCacheSize);
     LogPrintf("Using %zu MiB out of %zu requested for signature cache, able to store %zu elements\n",
```
Would be nice to update the LogPrintf here (and in the new Init) to say something about how the space is divided among multiple caches now, so users aren't confused why they aren't getting the full allocation here.

TheBlueMatt replied:
OK, done, also updated the doc for -maxsigcachesize.

TheBlueMatt committed:
Update -maxsigcachesize doc clarify init logprints for it

morcos commented:
Concept ACK
But I'd prefer if there were more safeguards in place against future changes that might cause consensus failure.  For instance, I think anything that is inputted to CScriptCheck should be committed to by the hash.  Right now it its the case that anything USED by CScriptCheck is committed to, but there is nothing stopping a future change to CScriptCheck that used the height from the Coins or something and caused problems.

jtimon commented on src/validation.cpp:
```diff
@@ -1669,6 +1748,41 @@ class WarningBitsConditionChecker : public AbstractThresholdConditionChecker
 // Protected by cs_main
 static ThresholdConditionCache warningcache[VERSIONBITS_NUM_BITS];
 
+static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const CChainParams& chainparams) {
```
This function can take Consensus::Params directly instead of CChainParams
Can we pass ```ThresholdConditionCache& cache``` to this and move it to src/versionbits ?
I know not all flags are activated with bip9, but it seems the best place to me.
Doing so would prevent you from calling IsWitnessEnabled() and AssertLockHeld(cs_main), but do you need to?
You can do the latter from the caller like we do with VersionBitsState

TheBlueMatt replied:
I'd much prefer to encapsulate the version bits stuff in GetBlockScriptFlags, and keep the diff on the smaller side. You're welcome to move things around afterwards, validation.cpp need to continue its trek towards better internal interfaces.

I changed it to take Consensus::Params.

jtimon replied:
Moving it to src/versionbits.cpp would still encapsulate version bits stuff in GetBlockScriptFlags, you would just not be using the versionbitscache and cs_main globals from validation.cpp.
You will also avoid using the unnecessary wrapper ```IsWitnessEnabled()``` (specially unnecessary in this function since you are already using ```VersionBitsState()``` directly for csv anyway).

It is "free" to do it now vs moving it in the future (if that ever gets the review). It may not be important to your approach to libconsensus, but it is to mine. I've been trying to do this for a long time, @CodeShark tried something similar as preparation to bip9 too.

TheBlueMatt replied:
I dont think it makes any sense to move this entire function into src/versionbits.cpp? This function has stuff for BIP 65, 66 and 16, none of which are versionbits-activated? Moving all soft forks into a separate thing is definitely a separate issue from this PR.

jtimon replied:
I noted that at the beginning "I know not all flags are activated with bip9", but I still think it's the best place for this new function (even if some lines of the function don't rely on versionbits, they also don't rely on anything that versionbits didn't already rely on).Doing a separate module only for this single function seems stupid to me.
Anyway, as disappointing as your rejection to move the new function for free it is to me, let's move on. 

Can you at least avoid the use of globals you don't need to use, ie versionbitscache and cs_main  (both used directly and indirectly through IsWitnessEnabled())?
That way, if we ever move this function out of validation (like all previous attempts at creating it did) it can be a clean moveonly commit (whether it goes to versionbits.o or somewhere else) instead of needing to change the function signature.

jtimon commented on src/validation.cpp:
```diff
@@ -451,6 +451,8 @@ int64_t GetTransactionSigOpCost(const CTransaction& tx, const CCoinsViewCache& i
 }
 
+// Returns the script flags which should be checked for a given block
+static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const CChainParams& chainparams);
```
why a separated declaration if it's going to be static?

JeremyRubin replied:
It's a forward declaration. As to why the definition isn't just there, I'm not sure, maybe GetBlockScriptFlags also has some dependency that would need a forward decl.

TheBlueMatt replied:
The reason its down further is that I didnt want to move it up with the mempool stuff in validation.cpp, but its called from ATMP, so needed the forward declaration. Hopefully this stuff gets ironed out as we move towards better internal interfaces (see #10279).

jtimon replied:
You are creating the function as new, it won't be any more disruptive to avoid the forward declaration. Anyway, my preference would be to move it to versionbits.o which is also ignored as "something that can be ironed out later", so whatever.

jtimon commented on src/validation.cpp:
history bike-sheeding currentBlockScriptVerifyFlags should be part of the previous commit, not this one.

jtimon commented on src/validation.cpp:
Perhaps everything that is below this case could be moved to CheckInputsFromMempoolAndCache too

TheBlueMatt commented on src/validation.cpp:
Its needed in this PR because its much more likely to trigger after the changes in this PR, though still unsupported. I'm happy to make it a PR on its own if you think its really worth it, but it seems to me this PR is still of very-reviewable-size.

TheBlueMatt commented on src/validation.cpp:
I dont think CheckInputsFromMempoolAndCache should make assumptions about having called CheckInputs already prior to this call? Maybe I'm confused as to what you're suggesting here.

TheBlueMatt commented:
@morcos I added an additional commit to only pass the scriptPubKey and nValue from the prevout into CScriptCheck, so hopefully any such future changes would be super clear to reviewers as consensus bugs. Sadly I dont really want to just include the height in the hash, as there are many heights, but I think this is a sufficient change.

jtimon commented on src/validation.cpp:
You misunderstood. Creating a currentBlockScriptVerifyFlags local variable just makes the next line simpler, which is fine, but could have been done in "Cache full script execution results in addition to signatures" directly which already touches those lines instead of "Do not print soft-fork-script warning with -promiscuousmempool", which would be clearer without that extra bikeshedding +2-1.
Anyway, as said, history bikeshedding, not a big deal

jtimon commented on src/validation.cpp:
I mean moving the following inside the function too: 
```
         {
              // If we're using promiscuousmempoolflags, we may hit this normally
              // Check if current block has some flags that scriptVerifyFlags
             // does not before printing an ominous warning
             if (!(~scriptVerifyFlags & currentBlockScriptVerifyFlags))
                 return error("%s: BUG! PLEASE REPORT THIS! ConnectInputs failed against MANDATORY but not STANDARD flags %s, %s",
                     __func__, hash.ToString(), FormatStateMessage(state));
         }
```
Then here you would just call:

```
if (!CheckInputsFromMempoolAndCache(tx, state, view, pool, scriptVerifyFlags, currentBlockScriptVerifyFlags, true, txdata)) {
   return false;
}

jtimon commented on src/validation.cpp:
The error should no be printed, but the function still needs to return false, no?

jtimon commented on src/validation.cpp:
Perhaps even put the 2 calls to checkinputs in it, from https://github.com/bitcoin/bitcoin/pull/10192/commits/061d06faceed0272d1913800369768188876df77#diff-24efdb00bfbe56b140fb006b562cc70bR889

TheBlueMatt committed:
Do not print soft-fork-script warning with -promiscuousmempool

TheBlueMatt commented on src/validation.cpp:
Fixed.

TheBlueMatt commented on src/validation.cpp:
But that logging implicitly makes an assumption about the fact that you've already called CheckInputs prior to the CheckInputsFromMempoolAndCache call, which I'd prefer not to do. CheckInputsFromMempoolAndCache *only* does some additional mempool-shouldnt-pollute-cache checks, and then calls CheckInputs.

TheBlueMatt commented on src/validation.cpp:
Fixed.

sdaftuar commented:
Needs rebase

gmaxwell commented:
@TheBlueMatt REBASE ME

TheBlueMatt commented:
Rebased :).

sipa commented:
Needs moar rebase.

sdaftuar commented on src/validation.cpp:
```diff
@@ -732,8 +732,20 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState& state, const C
+                    return error("%s: ConnectInputs failed against MANDATORY but not STANDARD flags due to promiscuous mempool %s, %s",
+                        __func__, hash.ToString(), FormatStateMessage(state));
+                } else {
+                    LogPrintf("Warning: -promisuousmempool flags set to not include currently enforced soft forks, this may break mining or otherwise cause instability!\n");
```
spelling nit: "promiscuous"

sdaftuar commented on src/validation.cpp:
```diff
@@ -732,8 +732,20 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState& state, const C
+            // Check if current block has some flags that scriptVerifyFlags
+            // does not before printing an ominous warning
+            if (!(~scriptVerifyFlags & currentBlockScriptVerifyFlags)) {
+                return error("%s: BUG! PLEASE REPORT THIS! ConnectInputs failed against MANDATORY but not STANDARD flags %s, %s",
```
This error message isn't quite right; at this point we haven't actually compared against the MANDATORY flags.

sdaftuar reviewed (commented):
Needs (simple) rebase, but code review ACK apart from a couple nits.  Will test and profile.

jtimon commented:
https://github.com/bitcoin/bitcoin/pull/10427 introduces GetScriptFlags like here but with some of my nits solved. If merged first, should make this a little bit smaller and simpler to review.
This needs rebase again.

jtimon commented on src/validation.cpp:
```diff
@@ -310,6 +310,9 @@ bool CheckSequenceLocks(const CTransaction &tx, int flags, LockPoints* lp, bool
 }
 
 
+// Returns the script flags which should be checked for a given block
```
Why this? if you want to use it between this and its deifinition, why not move the definition here directly?

TheBlueMatt replied:
It feels much more reasonable next to the consensus code, as it is really critical consensus logic, instead of mixed in with all the ATMP code.

jtimon replied:
I'm not following, in this commit this additional declaration (when nobody is calling it between the declaration and the definition) is completely redundant. The definition is close to ConnectBlock, also consensus code.

luke-jr commented:
This sounds like it would break (or at least complicate) CHECKBLOCKVERSION and possibly even CHECKBLOCKATHEIGHT...?

sdaftuar commented:
> This sounds like it would break (or at least complicate) CHECKBLOCKVERSION and possibly even CHECKBLOCKATHEIGHT...?

I'm skeptical that we'd ever adopt such context-dependent script flags (which strike me as a bad idea), but putting that aside: I benchmarked this patch on data from December 2016, and observed a 31% reduction in total block validation time, which is such a large performance improvement that I don't think this should be held up by a hypothetical type of consensus change, which we've not done before.  We can always revisit this performance optimization in the future if necessary.

sipa commented:
Needs rebase.

gmaxwell commented:
@TheBlueMatt needs rebase

TheBlueMatt committed:
Add CheckInputs wrapper CCoinsViewMemPool -> non-consensus-critical

This wraps CheckInputs in ATMP's cache-inputs call to check that
each scriptPubKey the CCoinsViewCache provides is the one which
was committed to by the input's transaction hash.

TheBlueMatt commented:
Addressed @sdaftuar's comments and rebased.

sipa commented:
This does not look like it's rebased on top of 10195?

TheBlueMatt commented:
@sipa Yes it was, though I fucked up the second-to-last commit. Fixed now.

gmaxwell commented:
@TheBlueMatt the tests timed out and need kicking.

ryanofsky commented on src/validation.cpp:
```diff
@@ -1167,6 +1188,21 @@ bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsVi
+            // We only use the first 19 bytes of nonce to avoid a second SHA
+            // round - giving us 19 + 32 + 4 = 55 bytes (+ 8 + 1 = 64)
+            static_assert(55 - sizeof(flags) - 32 >= 128/8, "Want at least 128 bits of nonce for script execution cache");
+            CSHA256().Write(scriptExecutionCacheNonce.begin(), 55 - sizeof(flags) - 32).Write(tx.GetWitnessHash().begin(), 32).Write((unsigned char*)&flags, sizeof(flags)).Finalize(hashCacheEntry.begin());
```
In commit "Cache full script execution results in addition to signatures"

Would it be more efficient to avoid this hashing and the cache lookup below when `cacheFullScriptStore` is false?

TheBlueMatt replied:
I believe actual block checking sets cacheFullScriptStore to false, so that's kinda when you want it to cache lookup :).

ryanofsky replied:
> I believe actual block checking sets cacheFullScriptStore to false, so that's kinda when you want it to cache lookup :)

I see, `cacheFullScriptStore == false` really means "check and erase from cache" not "don't use cache" like I was thinking. Could add a comment about this, though in retrospect I guess it's pretty obvious.

TheBlueMatt replied:
Well its the same thing that the sigcache itself does, though I added a comment to the top of ConnectInputs.

ryanofsky commented on src/validation.cpp:
```diff
@@ -1167,6 +1188,21 @@ bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsVi
+            static_assert(55 - sizeof(flags) - 32 >= 128/8, "Want at least 128 bits of nonce for script execution cache");
+            CSHA256().Write(scriptExecutionCacheNonce.begin(), 55 - sizeof(flags) - 32).Write(tx.GetWitnessHash().begin(), 32).Write((unsigned char*)&flags, sizeof(flags)).Finalize(hashCacheEntry.begin());
+            AssertLockHeld(cs_main); //TODO: Remove this requirement by making CuckooCache not require external locks
+            if (scriptExecutionCache.contains(hashCacheEntry, !cacheFullScriptStore)) {
```
In commit "Cache full script execution results in addition to signatures"

If `pvChecks` is non-null, is it ok to just return true here without populating it? If so, it'd be good to have a comment here saying why, because it doesn't seem completely obvious.

TheBlueMatt replied:
I'm confused, pvChecks is passed in, you dont populate, just push back your set of to-be-validated CScriptChecks?

ryanofsky replied:
> I'm confused, pvChecks is passed in, you dont populate, just push back your set of to-be-validated CScriptChecks?

I guess I wouldn't expect `cacheFullScriptStore` being set to true or false to have an effect on what gets returned in `pvChecks`, and that an interaction between the two parameters would be documented.

The specific case where it looks like there is an interaction is after calling `CheckInputs` once with `pvChecks = null` and a script where `check() == true` for all inputs.

If `CheckInputs` is called after that on the same script with `pvChecks != null`, it seems like `pvChecks` will be empty if `cacheFullScriptStore` was previously true, and populated if it was previously false.

TheBlueMatt replied:
Added a better comment.

TheBlueMatt commented:
Rebased.

TheBlueMatt committed:
Better document CheckInputs parameter meanings

ryanofsky commented on src/validation.cpp:
```diff
@@ -1216,8 +1216,15 @@ void InitScriptExecutionCache() {
+ * script checks which are not necessary (eg due to script execution cache hits) are, obviously,
+ * not pushed onto pvChecks/run.
+ *
+ * Setting cacheSigStore/cacheFullScriptStore to true will remove elements from the corresponding cache
```
In commit "Better document CheckInputs parameter meanings"

s/true/false I think. Thanks for the comments!

ryanofsky commented on src/validation.cpp:
```diff
@@ -399,6 +399,35 @@ void UpdateMempoolForReorg(DisconnectedBlockTransactions &disconnectpool, bool f
+        LOCK(pool.cs);
+        for (const CTxIn& txin : tx.vin) {
+            const Coin& coin = view.AccessCoin(txin.prevout);
+            if (coin.IsSpent()) return false;
```
In commit "Add CheckInputs wrapper CCoinsViewMemPool -> non-consensus-critical"

Could you add a comment explaining the return? It's not clear to me if this is supposed to be an optimization or a paranoid safety check. If it's meant to be the latter, maybe a way to make it more paranoid would be to do `if (coin.IsSpent()) has_spent_input = true` here and assert `!CheckInputs || !has_spent_input` at the end.

sipa replied:
I believe the code is correct. No entry was found in the cache, so validation should fail - it's not a consistency bug.

ryanofsky replied:
> I believe the code is correct. No entry was found in the cache, so validation should fail - it's not a consistency bug.

I also think the code is correct, just would like a comment saying why this is a return rather than a continue, or something that asserts against the CheckInputs() result. The whole function except this one line seems intended to do a bunch of checks and catch logic errors, while this line bypasses checks and seems to be doing some kind of optimization. AFAICT. Right now the intent is unclear.

TheBlueMatt replied:
Added.

sipa commented on src/validation.cpp:
```diff
@@ -784,8 +784,20 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
+            // Check if current block has some flags that scriptVerifyFlags
+            // does not before printing an ominous warning
+            if (!(~scriptVerifyFlags & currentBlockScriptVerifyFlags)) {
+                return error("%s: BUG! PLEASE REPORT THIS! ConnectInputs failed against latest-block but not STANDARD flags %s, %s",
```
No way to structure this so you can avoid the duplicate error line?

TheBlueMatt replied:
They arent duplicates? One prints that an issue has been found (and should be reported), the other simply warns you about your promiscuousmempool flags.

sipa commented on src/validation.cpp:
```diff
@@ -399,6 +399,35 @@ void UpdateMempoolForReorg(DisconnectedBlockTransactions &disconnectpool, bool f
+                assert(txFrom->GetHash() == txin.prevout.hash);
+                assert(txFrom->vout.size() > txin.prevout.n);
+                assert(txFrom->vout[txin.prevout.n] == coin.out);
+            } else {
```
I don't think this section is necessary. The goal of this function should just be checking that no mempool bugs are leaking into consensus validation. This branch looks like an unrelated sanity check.

TheBlueMatt replied:
The goal of this function includes checking CCoinsViewMempool, which is mempool code, from leaking into consensus code. As a part of that, this branch is required (and like the mempool get, this pcoinsTip lookup should be incredibly hot in cache, so it shouldn't be a big deal).

sipa commented on src/validation.cpp:
```diff
@@ -399,6 +399,35 @@ void UpdateMempoolForReorg(DisconnectedBlockTransactions &disconnectpool, bool f
+            if (coin.IsSpent()) return false;
+
+            const CTransactionRef& txFrom = pool.get(txin.prevout.hash);
+            if (txFrom) {
```
You can make the pool lookup conditional on `coin.nHeight == MEMPOOL_HEIGHT`. I think it's reasonable to rely on the fact the mempool - even if broken - will return Coin objects with nHeight set to that value.

TheBlueMatt replied:
I'd really prefer not to. The unordered_map lookup should essentially be free since anything required to find that transaction should already be very hot in cache, and if its free, why shouldnt we?

JeremyRubin commented:
Theoretical question: Have you considered making both caches store their hashes in the same instance of the data structure? Might work out a bit better as each can borrow unused capacity from one another.

TheBlueMatt commented:
Looks like it is to me?

On June 5, 2017 9:54:23 PM EDT, Pieter Wuille <notifications@github.com> wrote:
>This does not look like it's rebased on top of 10195?

sipa commented:
@TheBlueMatt Now it is. When I made the comment, this PR had `CCoins` in it in some places.

TheBlueMatt commented:
@JeremyRubin I have not, that would likely be interesting in the future (as well as possibly not making it an even 1/2-1/2 split in memory usage).

sipa commented on src/validation.cpp:
```diff
@@ -1666,7 +1779,7 @@ static bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockInd
             std::vector<CScriptCheck> vChecks;
             bool fCacheResults = fJustCheck; /* Don't cache results if we're actually connecting blocks (still consult the cache, though) */
-            if (!CheckInputs(tx, state, view, fScriptChecks, flags, fCacheResults, txdata[i], nScriptCheckThreads ? &vChecks : NULL))
+            if (!CheckInputs(tx, state, view, fScriptChecks, flags, fCacheResults, fCacheResults, txdata[i], nScriptCheckThreads ? &vChecks : NULL))
```
I think the second `fCacheResults` here can be `false`; as we normally pass a non-NULL `pvChecks` here, `fCacheFullScriptStore = true` has no effects anyway. This only affects `TestBlockValidity`.

TheBlueMatt replied:
fCacheFullScriptStore has a second meaning - it also deletes the element from the cache if a match is found, so we really should pass it through here to avoid deleting (or marking available) cache entries for TBV.

sipa replied:
Ah thanks, I missed that.

sipa commented:
utACK c435d9fbb76a5c4b5525ffdf8fc74afdac1a42df apart from nits

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -399,6 +399,40 @@ void UpdateMempoolForReorg(DisconnectedBlockTransactions &disconnectpool, bool f
+
+    assert(!tx.IsCoinBase());
+    {
+        LOCK(pool.cs);
```
I moved the lock up in the function and added a comment noting why we might want it (note that its already held outside of this function before calling it in ATMP).

sipa replied:
I'm still unconvinced it is needed, as it does not operate on CCoinsViewMemPool directly, but on a dissociated cache with the mempool UTXO loaded into it. If it's valid at one point with those inputs, it's also valid later.

sdaftuar committed:
Add CheckInputs() unit tests

Check that cached script execution results are only valid for the same
script flags; that script execution checks are returned for non-cached
transactions; and that cached results are only valid for transactions
with the same witness hash.

TheBlueMatt commented:
@sdaftuar pointed out that we could directly test CheckInputs' use of its own cache in unit tests, so I added a rather simple one that did so.

sipa commented on src/test/checkinputs_script_cache_tests.cpp:
```diff
@@ -0,0 +1,69 @@
+    LOCK(cs_main);
+
+    CValidationState state;
+    PrecomputedTransactionData txdata(spends[0]);
```
spends is undeclared

TheBlueMatt replied:
Heh, thats what happens when you change a variable type after fixing the test.

sipa commented:
utACK 316d328b915f2e23bb750096ee9878b14dfd0a26

gmaxwell reviewed (approved):
ACK.   It's worth pointing out that we have no system test that fails when the flags are mishandled here-- though the unit test correctly fails.

TheBlueMatt commented:
I believe @sdaftuar indicated that he had a better test written for this, not sure what format it takes, however.

sipa commented:
Verified empirically that this actually gives a performance improvement:

Last 10 block verifications on my server (benchmarked using -debug=bench):

On master as of a few weeks ago:
```
2017-06-26 22:16:21.499109     - Verify 5160 txins: 118.87ms (0.023ms/txin) [96.86s]
2017-06-26 22:22:34.710518     - Verify 3957 txins: 84.70ms (0.021ms/txin) [96.95s]
2017-06-26 22:32:43.025738     - Verify 4333 txins: 85.08ms (0.020ms/txin) [97.03s]
2017-06-26 22:41:47.084320     - Verify 4300 txins: 83.29ms (0.019ms/txin) [97.12s]
2017-06-26 22:46:33.711967     - Verify 4650 txins: 60.67ms (0.013ms/txin) [97.18s]
2017-06-26 22:53:00.210314     - Verify 4656 txins: 61.19ms (0.013ms/txin) [97.24s]
2017-06-26 22:54:35.550521     - Verify 5128 txins: 100.05ms (0.020ms/txin) [97.34s]
2017-06-26 23:16:50.156286     - Verify 4889 txins: 118.09ms (0.024ms/txin) [97.46s]
2017-06-26 23:18:17.993009     - Verify 4973 txins: 104.94ms (0.021ms/txin) [97.56s]
2017-06-26 23:27:55.039898     - Verify 4712 txins: 92.86ms (0.020ms/txin) [97.65s]
```

After restarting with this PR:
```
2017-06-26 23:38:16.352833     - Verify 6016 txins: 39.89ms (0.007ms/txin) [0.26s]
2017-06-26 23:38:37.499180     - Verify 6650 txins: 33.67ms (0.005ms/txin) [0.29s]
2017-06-26 23:52:52.539748     - Verify 4686 txins: 60.94ms (0.013ms/txin) [0.35s]
2017-06-27 00:34:42.788390     - Verify 4417 txins: 61.53ms (0.014ms/txin) [0.42s]
2017-06-27 00:38:47.022972     - Verify 3023 txins: 33.88ms (0.011ms/txin) [0.45s]
2017-06-27 00:42:59.372504     - Verify 4289 txins: 38.45ms (0.009ms/txin) [0.49s]
2017-06-27 00:44:06.469073     - Verify 5065 txins: 34.92ms (0.007ms/txin) [0.52s]
2017-06-27 00:53:41.905643     - Verify 4378 txins: 72.81ms (0.017ms/txin) [0.60s]
2017-06-27 01:04:23.953806     - Verify 4917 txins: 52.77ms (0.011ms/txin) [0.65s]
2017-06-27 01:08:59.598035     - Verify 5002 txins: 48.47ms (0.010ms/txin) [0.70s]
```

TheBlueMatt commented:
Replaced test with one by @sdaftuar which is much better. Should be good to go now.

sdaftuar commented:
ACK e3f9c05

sipa commented on src/test/txvalidationcache_tests.cpp:
```diff
@@ -84,4 +90,282 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_block_doublespend, TestChain100Setup)
+        // CheckInputs should succeed iff test_flags doesn't intersect with
+        // failing_flags
+        bool expected_return_value = !(test_flags & failing_flags);
+        if (expected_return_value && upgraded_nop) {
```
This logic may become unnecessary with #10699.

gmaxwell reviewed (commented):
reACK w/ new tests

morcos commented:
posthumous utACK

Thanks for doing this.

JeremyRubin commented:
probably wrong place to have this conversation, but for the sake of continuity...

> @JeremyRubin I have not, that would likely be interesting in the future (as well as possibly not making it an even 1/2-1/2 split in memory usage).

Yeah I've (asynchronously) been thinking about this one a little bit in spare cycles. There are a couple tricks on could do to easily increase frequency, e.g. add a insert_n_times to the cuckoocache where you can specify that it should be put onto 1-8 of it's hash locations. This makes the hash "stickier" in memory compared to other entries and is pretty easy to compute.

Generally, this also provides a really interesting mechanism for further tuning the hit rate based on other priors. E.g., if a txn comes in and we estimate that it has a 0.5 chance of entering a block, we could set it to have, say, half of it's hash values inserted onto. If a txn comes in and we estimate that it has 0.25 we could fill a quarter of them.

This can also be simulated by inserting the hash N times with different salts or something without modifying the cuckoocache internally.
